<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Case study - nesting • rpharma2020tidytranscriptomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Case study - nesting">
<meta property="og:description" content="rpharma2020tidytranscriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rpharma2020tidytranscriptomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/tidytranscriptomics.html">Workshop</a>
</li>
<li>
  <a href="../articles/supplementary.html">Supplementary</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/rpharma2020_tidytranscriptomics">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/repository/docker/stemangiola/rpharma2020_tidytranscriptomics">
    <span class="fa fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="nesting_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Case study - nesting</h1>
                        <h4 class="author">Maria Doyle, Peter MacCallum Cancer Centre<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 class="author">Stefano Mangiola, Walter and Eliza Hall Institute<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/rpharma2020_tidytranscriptomics/blob/master/vignettes/nesting.Rmd"><code>vignettes/nesting.Rmd</code></a></small>
      <div class="hidden name"><code>nesting.Rmd</code></div>

    </div>

    
    
<p><img src="../inst/vignettes/tidybulk_logo.png" width="200px"></p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidybulk</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'tidybulk'</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     arrange, bind_cols, bind_rows, distinct, filter, full_join,</span>
<span class="co">#&gt;     group_by, inner_join, left_join, mutate, rename, right_join,</span>
<span class="co">#&gt;     rowwise, summarise, ungroup</span>
<span class="co">#&gt; The following object is masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter</span></pre></div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><code>tidybulk</code> allows for data nesting, using the <code>tidyr</code> utility <code>nest</code>. This is an extremely powerful tool as allows functional programming on transcriptomic data.</p>
<div id="nesting-for-performing-the-same-analyses-independently-for-subsets" class="section level3">
<h3 class="hasAnchor">
<a href="#nesting-for-performing-the-same-analyses-independently-for-subsets" class="anchor"></a>Nesting for performing the same analyses independently for subsets</h3>
<p>Let’s suppose we want to perform differential transcript abundance analysis independently for two different data subsets to compare results after the test</p>
<div class="sourceCode" id="cb2"><pre class="downlit">

<span class="va">pasilla_de</span> <span class="op">=</span> 
  <span class="fu">rpharma2020tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/pasilla.html">pasilla</a></span> <span class="op">%&gt;%</span> 
  
  <span class="co"># Convert SE object to tibble</span>
  <span class="va">tidybulk</span> <span class="op">%&gt;%</span>
  
   <span class="co"># Scale abundance for plotting</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest<span class="op">=</span><span class="va">condition</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Nest</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Differential analysis</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span>
    <span class="va">data</span>,
    <span class="op">~</span>  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span><span class="va">.x</span>, <span class="op">~</span> <span class="va">condition</span><span class="op">)</span> 
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidyr-methods.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>
<span class="co">#&gt; tidybulk says: All methods use raw counts, </span>
<span class="co">#&gt; irrespective of if scale_abundance or adjust_abundance have been calculated, </span>
<span class="co">#&gt; therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; tidybulk says: The design column names are "(Intercept), conditionuntreated"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span>
<span class="co">#&gt; tidybulk says: All methods use raw counts, </span>
<span class="co">#&gt; irrespective of if scale_abundance or adjust_abundance have been calculated, </span>
<span class="co">#&gt; therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; tidybulk says: The design column names are "(Intercept), conditionuntreated"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span></pre></div>
<p>Now we can for example compare the number of differentially transcribed genes and their co-expression</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">pasilla_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>
    number_of_differential <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_int</a></span><span class="op">(</span>
      <span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">nrow</span>
    <span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 2 x 3</span>
<span class="co">#&gt;   type       data                   number_of_differential</span>
<span class="co">#&gt;   &lt;fct&gt;      &lt;list&gt;                                  &lt;int&gt;</span>
<span class="co">#&gt; 1 single_end &lt;tibble [43,797 × 13]&gt;                     40</span>
<span class="co">#&gt; 2 paired_end &lt;tibble [58,396 × 13]&gt;                   1270</span></pre></div>
<p>We can easily see which genes overlap, and plot them</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">pasilla_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">feature</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>occurrences <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_int</a></span><span class="op">(</span><span class="va">data</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">nrow</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># We filter some of them</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">occurrences</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidyr-methods.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># And plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">type</span>, <span class="va">counts_scaled</span> <span class="op">+</span><span class="fl">1</span>, color<span class="op">=</span><span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">feature</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p><img src="nesting_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="nesting-for-applying-to-the-full-data-frame-analyses-done-of-part-of-the-data-frame" class="section level3">
<h3 class="hasAnchor">
<a href="#nesting-for-applying-to-the-full-data-frame-analyses-done-of-part-of-the-data-frame" class="anchor"></a>Nesting for applying to the full data frame analyses done of part of the data frame</h3>
<p>Let’s suppose we want to indentify the markers that distingush epithelial from endothelial cells, but we want to then visualising those genes across many cell types to understand the background transcription.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">

<span class="va">cell_type_tt</span> <span class="op">=</span> <span class="fu">rpharma2020tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/cell_type_df.html">cell_type_df</a></span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">symbol</span>, <span class="va">count</span><span class="op">)</span>

<span class="va">markers_df</span> <span class="op">=</span> 
  <span class="va">cell_type_tt</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Scale abundance for plotting</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest<span class="op">=</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Nest</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># investigate one cell type pair</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>comparison_data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span>
    <span class="va">data</span>,
    <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">cell_type</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"endothelial"</span>, <span class="st">"epithelial"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cell_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">)</span> 
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co">#test. We run on the two populations but we select data for all populations</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>markers <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span>
    <span class="va">comparison_data</span>,
    <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span>
      
      <span class="co"># Differential transcription</span>
      <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">cell_type</span>, 
        .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cell_typeendothelial - cell_typeepithelial"</span><span class="op">)</span>,
        action<span class="op">=</span><span class="st">"only"</span>, 
        omit_contrast_in_colnames <span class="op">=</span> <span class="cn">TRUE</span>
      <span class="op">)</span> <span class="op">%&gt;%</span>
      
      <span class="co"># Select markers</span>
      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span> 
      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span> 
      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>
    
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Add marker info to original data</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html">map2</a></span><span class="op">(</span><span class="va">data</span>, <span class="va">markers</span>, <span class="op">~</span> <span class="va">.x</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">%in%</span> <span class="va">.y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidyr-methods.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> 
<span class="co">#&gt; tidybulk says: All methods use raw counts, </span>
<span class="co">#&gt; irrespective of if scale_abundance or adjust_abundance have been calculated, </span>
<span class="co">#&gt; therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; tidybulk says: The design column names are "cell_typeendothelial, cell_typeepithelial"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span></pre></div>
<p>Now we can plot our markers including the background</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">markers_df</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">cell_type</span>, <span class="va">count_scaled</span> <span class="op">+</span><span class="fl">1</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p><img src="nesting_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div id="contributing" class="section level2">
<h2 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h2>
<p>If you want to suggest improvements for this workshop or ask questions, you can do so as described <a href="https://github.com/stemangiola/tidytranscriptomics/blob/master/CONTRIBUTING.md">here</a>.</p>
</div>
<div id="reproducibility" class="section level2">
<h2 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h2>
<p>Record package and version information with <code>sessionInfo</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.2 Patched (2020-09-24 r79253)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.1 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] tidybulk_1.1.8 purrr_0.3.4    ggplot2_3.3.2  dplyr_1.0.2   </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.5                           locfit_1.5-9.4                      </span>
<span class="co">#&gt;  [3] lattice_0.20-41                      tidyr_1.1.2                         </span>
<span class="co">#&gt;  [5] utf8_1.1.4                           assertthat_0.2.1                    </span>
<span class="co">#&gt;  [7] rprojroot_1.3-2                      digest_0.6.25                       </span>
<span class="co">#&gt;  [9] R6_2.4.1                             GenomeInfoDb_1.25.11                </span>
<span class="co">#&gt; [11] backports_1.1.10                     stats4_4.0.2                        </span>
<span class="co">#&gt; [13] evaluate_0.14                        pillar_1.4.6                        </span>
<span class="co">#&gt; [15] zlibbioc_1.35.0                      rlang_0.4.7                         </span>
<span class="co">#&gt; [17] S4Vectors_0.27.13                    Matrix_1.2-18                       </span>
<span class="co">#&gt; [19] preprocessCore_1.51.0                rmarkdown_2.3                       </span>
<span class="co">#&gt; [21] pkgdown_1.6.1                        desc_1.2.0                          </span>
<span class="co">#&gt; [23] splines_4.0.2                        readr_1.3.1                         </span>
<span class="co">#&gt; [25] stringr_1.4.0                        RCurl_1.98-1.2                      </span>
<span class="co">#&gt; [27] munsell_0.5.0                        DelayedArray_0.15.9                 </span>
<span class="co">#&gt; [29] compiler_4.0.2                       xfun_0.17                           </span>
<span class="co">#&gt; [31] pkgconfig_2.0.3                      systemfonts_0.3.1                   </span>
<span class="co">#&gt; [33] BiocGenerics_0.35.4                  htmltools_0.5.0                     </span>
<span class="co">#&gt; [35] tidyselect_1.1.0                     SummarizedExperiment_1.19.6         </span>
<span class="co">#&gt; [37] tibble_3.0.3                         GenomeInfoDbData_1.2.3              </span>
<span class="co">#&gt; [39] edgeR_3.31.4                         IRanges_2.23.10                     </span>
<span class="co">#&gt; [41] matrixStats_0.57.0                   fansi_0.4.1                         </span>
<span class="co">#&gt; [43] crayon_1.3.4                         withr_2.3.0                         </span>
<span class="co">#&gt; [45] bitops_1.0-6                         grid_4.0.2                          </span>
<span class="co">#&gt; [47] gtable_0.3.0                         lifecycle_0.2.0                     </span>
<span class="co">#&gt; [49] magrittr_1.5                         rpharma2020tidytranscriptomics_1.0.6</span>
<span class="co">#&gt; [51] scales_1.1.1                         cli_2.0.2                           </span>
<span class="co">#&gt; [53] stringi_1.5.3                        farver_2.0.3                        </span>
<span class="co">#&gt; [55] XVector_0.29.3                       fs_1.5.0                            </span>
<span class="co">#&gt; [57] limma_3.45.14                        ellipsis_0.3.1                      </span>
<span class="co">#&gt; [59] ragg_0.3.1                           generics_0.0.2                      </span>
<span class="co">#&gt; [61] vctrs_0.3.4                          tools_4.0.2                         </span>
<span class="co">#&gt; [63] Biobase_2.49.1                       glue_1.4.2                          </span>
<span class="co">#&gt; [65] hms_0.5.3                            parallel_4.0.2                      </span>
<span class="co">#&gt; [67] yaml_2.2.1                           colorspace_1.4-1                    </span>
<span class="co">#&gt; [69] GenomicRanges_1.41.6                 cpp11_0.2.1                         </span>
<span class="co">#&gt; [71] memoise_1.1.0                        knitr_1.30</span></pre></div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:maria.doyle@petermac.org" class="email">maria.doyle@petermac.org</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p><a href="mailto:mangiola.s@wehi.edu.au" class="email">mangiola.s@wehi.edu.au</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
